### PR TITLE
ETL-2565: Error handling in camus for LeaderNotAvailable in TopicMetadata

### PR DESCRIPTION
**Pull request by ydailinkedin**
Addressed the review comments. Added unit test.

### CODE REVIEW

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 1]
```diff
@@ -461,6 +464,17 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
     }
     return offsetKeysMap;
   }
+  
+  private void refreshPartitionMetadata(PartitionMetadata partitionMetadata, TopicMetadata topicMetadata, JobContext context) {
+    List<TopicMetadata> topicMetadataList = this.getKafkaMetadata(context, Collections.singletonList(topicMetadata.topic()));
+    topicMetadata = topicMetadataList.get(0);
+    for(PartitionMetadata metadataPerPartition : topicMetadata.partitionsMetadata()) {
+      if (metadataPerPartition.partitionId() == partitionMetadata.partitionId()) {
+        partitionMetadata = metadataPerPartition;
```
Comment (zliu41) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: This has no effect to partitionMetadata in line 265, right? java is pass by value.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 2]
```diff
@@ -263,6 +262,10 @@ public String createTopicRegEx(HashSet<String> topicsSet) {
             // We only care about LeaderNotAvailableCode error on partitionMetadata level
             // Error codes such as ReplicaNotAvailableCode should not stop us.
             if (partitionMetadata.errorCode() == ErrorMapping.LeaderNotAvailableCode()) {
+              partitionMetadata = refreshPartitionMetadata(partitionMetadata, topicMetadata, context);
```
Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Add a logging before retry

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 3]
```diff
@@ -461,6 +464,17 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
     }
     return offsetKeysMap;
   }
+  
+  private PartitionMetadata refreshPartitionMetadata(PartitionMetadata partitionMetadata, TopicMetadata topicMetadata, JobContext context) {
+    List<TopicMetadata> topicMetadataList = this.getKafkaMetadata(context, Collections.singletonList(topicMetadata.topic()));
+    topicMetadata = topicMetadataList.get(0);
```
Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Should check Exception and null pointer/size check on topicMetadataList

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 4]
```diff
@@ -461,6 +464,17 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
...
+    for(PartitionMetadata metadataPerPartition : topicMetadata.partitionsMetadata()) {
+      if (metadataPerPartition.partitionId() == partitionMetadata.partitionId()) {
+        return metadataPerPartition;
+      }
+    }
```
Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: If the loop finishes but no partition matches, it should be treated as a different error case.

And how did you test your code?

We also need to add unit test.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 5]
```diff
@@ -99,8 +99,7 @@ public static void setLogger(Logger log) {
    * @param context
    * @return
    */
-  public List<TopicMetadata> getKafkaMetadata(JobContext context) {
-    ArrayList<String> metaRequestTopics = new ArrayList<String>();
+  public List<TopicMetadata> getKafkaMetadata(JobContext context, List<String> metaRequestTopics) {
```
Comment (chavdar) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Please add a javadoc for the metaRequestTopics. For example, that empty list means get all topics.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java (hunk 2 above)
Comment (chavdar) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Shouldn't this be in a while loop for multiple retries with some backoff?

Comment (ydailinkedin) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: I will add a constant for retrying multiple times. But I am not sure if the backoff is necessary here. The method getKafkaMetadata contacts the all the brokers in the re-shuffled broker list for the topicMetadata.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 6]
```diff
@@ -527,6 +535,39 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
     return offsetKeysMap;
   }
 
+  public PartitionMetadata refreshPartitionMetadataOnLeaderNotAvailable(PartitionMetadata partitionMetadata,
+      TopicMetadata topicMetadata, JobContext context, int retryTimes) throws InterruptedException {
+    int retryCounter = 0;
+    while (partitionMetadata.errorCode() == ErrorMapping.LeaderNotAvailableCode() && retryCounter < retryTimes) {
+      log.info("Retry to referesh the topicMetadata on LeaderNotAvailable...");
+      Thread.sleep((retryCounter + 1) * BACKOFF_UNIT_MILLISECONDS);
```
Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: We usually handles InterruptedException here.

Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: You can also wait for a Random number of milliseconds.

Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: The sleep should happen after the 1st try, not before.

File: camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormatTest.java [hunk 7]
```diff
@@ -94,11 +103,63 @@ public void testWithThreeRetries() {
...
+        etlInputFormat.refreshPartitionMetadataOnLeaderNotAvailable(partitionMetadata2, mockedTopicMetadata,
+            dummyContext, 1));
+    
+    //For partitionMetadata1, it will refresh if the errorcode is LeaderNotAvailable.
+    assertEquals(
+        mockedReturnedPartitionMetadata,
+        etlInputFormat.refreshPartitionMetadataOnLeaderNotAvailable(partitionMetadata1, mockedTopicMetadata,
+            dummyContext, 1));
+  }
+
```
Comment (HenryCaiHaiying) [inline camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormatTest.java]: Maybe another test case on all retries failed?

Comment (ydailinkedin) [inline camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormatTest.java]: New test case added.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java (hunk 6 above)
Comment (ydailinkedin) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Changed.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 8]
```diff
@@ -80,6 +80,9 @@
   public static final String CAMUS_WORK_ALLOCATOR_CLASS = "camus.work.allocator.class";
   public static final String CAMUS_WORK_ALLOCATOR_DEFAULT = "com.linkedin.camus.workallocater.BaseAllocator";
 
+  public static final int RETRY_TIMES = 3;
```
Comment (zliu41) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Use similar names as line 86 & 87 (NUM_TRIES_xxx)

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 9]
```diff
@@ -527,6 +533,39 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
...
+      } else {
+        topicMetadata = topicMetadataList.get(0);
+        boolean partitionFound = false;
+        for (PartitionMetadata metadataPerPartition : topicMetadata.partitionsMetadata()) {
+          if (metadataPerPartition.partitionId() == partitionMetadata.partitionId()) {
+            partitionFound = true;
+            if (metadataPerPartition.errorCode() != ErrorMapping.LeaderNotAvailableCode()) {
+              return metadataPerPartition;
+            } else { //retry again.
+              Thread.sleep((long) (Math.random() * (retryCounter + 1) * BACKOFF_UNIT_MILLISECONDS));
```
Comment (zliu41) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: Shouldn't sleep after the last try.

File: camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java [hunk 10]
```diff
@@ -527,6 +533,39 @@ private void writeRequests(List<CamusRequest> requests, JobContext context) thro
     return offsetKeysMap;
   }
 
+  public PartitionMetadata refreshPartitionMetadataOnLeaderNotAvailable(PartitionMetadata partitionMetadata,
+      TopicMetadata topicMetadata, JobContext context, int retryTimes) throws InterruptedException {
```
Comment (zliu41) [inline camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java]: "retryTimes" is ambiguous. Does it include or exclude the first try? Better to use "tryTimes" or "numTries"

**Comment by zliu41** [comment]
Please: (1) run a regression test on nertz; (2) pull from upstream and resolve conflict.

**Comment by ydailinkedin** [comment]
The regression test succeeds on Nertz.

**Pull request merged**